datalad/support/gitrepo.py

```diff
@@ -2362,15 +2370,115 @@ def gc(self, allow_background=False, auto=False):
             cmd_options += ['--auto']
         self._git_custom_command('', cmd_options)
 
-    def get_submodules(self, sorted_=True):
-        """Return a list of git.Submodule instances for all submodules"""
-        # check whether we have anything in the repo. if not go home early
-        if not self.repo.head.is_valid():
-            return []
-        submodules = self.repo.submodules
+    def _parse_gitmodules(self):
+        # TODO read .gitconfig from Git blob?
+        gitmodules = self.pathobj / '.gitmodules'
+        if not gitmodules.exists():
+            return {}
+        # pull out file content
+        out, err = self._git_custom_command(
+            '',
+            ['git', 'config', '-z', '-l', '--file', '.gitmodules'])
+        # abuse our config parser
+        db, _ = _parse_gitconfig_dump(out, {}, None, True)
+        mods = {}
+        for k, v in iteritems(db):
+            if not k.startswith('submodule.'):
+                # we don't know what this is
+                lgr.debug("Skip unrecognized .gitmodule specification: %s=%s", k, v)
+                continue
+            k_l = k.split('.')
+            # module name is everything after 'submodule.' that is not the variable
+            # name
+            mod_name = '.'.join(k_l[1:-1])
+            mod = mods.get(mod_name, {})
+            # variable name is the last 'dot-free' segment in the key
+            mod[k_l[-1]] = v
+            mods[mod_name] = mod
+
+        out = {}
+        # bring into traditional shape
+        for name, props in iteritems(mods):
+            if 'path' not in props:
+                lgr.debug("Failed to get '%s.path', skipping section", name)
+                continue
+            modprops = {'gitmodule_{}'.format(k): v
+                        for k, v in iteritems(props)
+                        if not (k.startswith('__') or k == 'path')}
+            modpath = self.pathobj / PurePosixPath(props['path'])
+            modprops['gitmodule_name'] = name
+            out[modpath] = modprops
+        return out
+
+    def get_submodules_(self, paths=None):
+        """Yield submodules in this repository.
+
+        Parameters
+        ----------
+        paths : list(pathlib.PurePath), optional
+            Restrict submodules to those under `paths`.
+
+        Returns
+        -------
+        A generator that yields a dictionary with information for each
+        submodule.
+        """
+        if not ((self.pathobj / ".gitmodules").exists() and self.get_hexsha()):
```

Comment:
Re `self.get_hexsha()`...  With this changes we will be returning submodules are known to `.gitmodules`, either dirty or not . Restricting by demanding also having a commit, although very unlikely and impossible when working with datalad datasets, seems to be not necessary.  Or what am I missing?

Reply:
> Restricting by demanding also having a commit, although very unlikely and impossible when working with datalad datasets, seems to be not necessary.

Good point.  It's not necessary.  I was porting the `self.repo.head.is_valid()` check from `GitRepo.get_submodules()`, but that indeed changes the behavior of `datalad subdataset` when there is no commit checked out.  So `GitRepo.get_submodules()` and `subdatasets()` differ in their treatment of this case, and for the rewrite of `get_submodules()` I agree we should use the less restrictive case.  Updated.

<details>
<summary>range-diff: v2 vs v3</summary>

```
 1:  ebadb1c42 =  1:  ebadb1c42 CLN: subdatasets: Drop unused imports
 -:  --------- >  2:  71e5c8ea6 TST: subdatasets: Test for "no commit checked out" case
 2:  23f30ba3c !  3:  7e9940945 MV: gitrepo: Absorb custom submodule parser from subdatasets()
    @@ -170,7 +170,7 @@
     +        return out
     +
     +    def get_submodules_(self, paths=None):
    -+        if not ((self.pathobj / ".gitmodules").exists() and self.get_hexsha()):
    ++        if not (self.pathobj / ".gitmodules").exists():
     +            return
     +
     +        modinfo = self._parse_gitmodules()
 3:  b488de78c !  4:  1f8b99dff DOC: gitrepo: Add docstring for get_submodules_()
    @@ -21,6 +21,6 @@
     +        A generator that yields a dictionary with information for each
     +        submodule.
     +        """
    -         if not ((self.pathobj / ".gitmodules").exists() and self.get_hexsha()):
    +         if not (self.pathobj / ".gitmodules").exists():
                  return
      
 4:  9392e2113 !  5:  bea8551d7 RF: gitrepo: Rewrite get_submodules() to avoid GitPython
    @@ -6,6 +6,13 @@
         GitPython, this avoids an unresolved issue in GitPython's handling of
         submodules [0].
     
    +    As documented by the new test, there is a change in behavior when
    +    get_submodules() is called in a repository that doesn't have a commit
    +    checked out: it now returns any registered submodules instead of an
    +    empty list.  This is consistent with the behavior of 'git submodule'
    +    and 'datalad subdataset', and there doesn't seem to be an obvious
    +    reason not to support it.
    +
         [0]: https://github.com/datalad/datalad/pull/3508#issuecomment-508462763
     
      diff --git a/datalad/support/gitrepo.py b/datalad/support/gitrepo.py
    @@ -86,3 +93,24 @@
      
          def is_submodule_modified(self, name, options=[]):
              """Whether a submodule has new commits
    +
    + diff --git a/datalad/support/tests/test_gitrepo.py b/datalad/support/tests/test_gitrepo.py
    + --- a/datalad/support/tests/test_gitrepo.py
    + +++ b/datalad/support/tests/test_gitrepo.py
    +@@
    +     raise SkipTest("TODO")
    + 
    + 
    ++@with_tempfile
    ++def test_get_submodules_parent_on_unborn_branch(path):
    ++    repo = GitRepo(path, create=True)
    ++    subrepo = GitRepo(op.join(path, "sub"), create=True)
    ++    subrepo.commit(msg="s", options=["--allow-empty"])
    ++    repo.add_submodule(path="sub")
    ++    eq_([s.name for s in repo.get_submodules()],
    ++        ["sub"])
    ++
    ++
    + def test_kwargs_to_options():
    + 
    +     class Some(object):
 5:  52737fdea =  6:  60f889d7d RF: request status only on provided paths (if given)
 6:  dba239dec =  7:  5fefda437 NF: get_parent_paths to be able to quickly determine within repo paths if paths within submodules provided
 7:  ad93ab7d6 =  8:  f8c41e0b8 BM: benchmark suite for get_parent_paths
 8:  fa14633c2 =  9:  bfa08474c BF: make use of get_parent_paths to mitigate difference in ls-tree and ls-files behavior on paths within submodules
 9:  d64ef83e7 = 10:  e7f4578de DOC: Use public-inbox for Git mailing list link
10:  c7905e74d = 11:  5a77f6ba2 OPT: do not sort, maintain a set of prior hits
```

</details>